llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -3476,7 +3469,7 @@ Instruction *InstCombinerImpl::visitOr(BinaryOperator &I) {
   if (Instruction *X = foldComplexAndOrPatterns(I, Builder))
     return X;
 
-  // (A == c & B != d) | (A != c & B == d)) -> (A == c) | (B == d)
+  // (A == c & B != d) | (A != c & B == d)) -> (A == c) ^ (B == d)
```

Comment:
very nit: can you be consistent on capitolization?

Reply:
Also, I guess can you generalize the comment, its really '(A & B) | (C & D) where A == ~C && B == ~D or A == ~D and B == ~D'

Finally, I think you can remove: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp#L1843 w/ this patch.

Reply:
removing it made the following test fail 
```
define i32 @and_to_xor4(i32 %a, i32 %b) {
; CHECK-LABEL: @and_to_xor4(
; CHECK-NEXT:    [[AND2:%.*]] = xor i32 [[B:%.*]], [[A:%.*]]
; CHECK-NEXT:    ret i32 [[AND2]]
;
  %or = or i32 %b, %a
  %and = and i32 %a, %b
  %not = xor i32 %and, -1
  %and2 = and i32 %not, %or
  ret i32 %and2
}

```

probably I need to move where I call my function to be called there as well so I make sure any negations are pushed inside

Reply:
What is the new result? Is it just `xor i32 %a, %b`? Or a regression?

Reply:
But yeah, to avoid any regressions you could probably just place the function call where the old impl was.

Reply:
to be able to do that we need to generalise `isKnownInversion` to handle the case of a  and ~a being inversions of each other. I will change that once we land this

Reply:
Oh wow, had just assumed `isKnownInversion` handled that case.

Personally would just do it all in this PR, but up to you.

The code as is now LGTM.

Reply:
> Oh wow, had just assumed `isKnownInversion` handled that case.
> 
> Personally would just do it all in this PR, but up to you.
> 
> The code as is now LGTM.

You told me not to handle this pattern :)
See https://github.com/llvm/llvm-project/pull/93591#issuecomment-2137997343

Reply:
I lack foresight :)

Reply:
Separate PR please.